api/src/pcapi/scripts/deposit/update_deposit_dates.py

```diff
@@ -0,0 +1,85 @@
+"""
+Le 13 avril vers 9h40, un script a été éxécuté pour mettre à jour les dates d'expiration des deposit.
+Dans ce script, il y avait une erreur :
+On regardait si la date contenait 0 pour les heures et les minutes, auquel cas on les transformait en 23 et 59.
+Pour les dates d'expiration qui étaient déjà au jour d'après, ça a repoussé le délai d'expiration de 23h59 :
+2021-31-05 10:49 sur la date d'OBTENTION donnait 2023-06-01 00:00 sur la date d'EXPIRATION et est passé à 2023-06-01 23:59.
+
+Ces heures étant UTC, sur le front actuellement (2023-05-23), un `new Date(expirationDate)` donne 2023-06-02 01:59,
+ce qui conduit à l'affichage 02/06/2023.
+
+De plus, pour une raison inconnue, l'éxécution du script a été apparemment interrompue à la moitié, laissant en base
+des dates d'expiration à 00:00 du jour suivant et d'autres à 23:59 du jour suivant.
+
+Le but de ce script ici est de toutes les passer à 23:59 de jour-même de l'expiration.
+"""
+
+import datetime
+from functools import wraps
+import logging
+import time
+
+from dateutil.relativedelta import relativedelta
+
+from pcapi.core.finance.models import Deposit
+from pcapi.models import db
+
+
+FAIL_FAST = True  # Use 'assert' instead of 'if ...: print'
+
+logging.basicConfig()
+logger = logging.getLogger("my-logger")
+logger.setLevel(logging.DEBUG)
+
+
+def timed(func):  # type: ignore[no-untyped-def]
+    """This decorator prints the execution time for the decorated function."""
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):  # type: ignore[no-untyped-def]
+        start = time.time()
+        result = func(*args, **kwargs)
+        end = time.time()
+        logger.debug("%s ran in %ss", func.__name__, round(end - start, 2))
+        return result
+
+    return wrapper
+
+
+@timed
+def update_deposit_expiration_dates(dry_run: bool = True, batch_size: int = 1000) -> None:
+    query = Deposit.query.filter(
+        Deposit.expirationDate > datetime.datetime.utcnow(),
+        Deposit.type == "GRANT_18",
+    )
+    total = query.count()
```

Comment:
Le nombre d'éléments est égal au plus grand identifiant... si et seulement s'il n'y a pas de trou dans la numérotation. Or cela peut arriver si on supprime un élément, ou si la séquence utilisée par PostgreSQL contient des trous (ce qui est le cas sur plusieurs tables, et probablement toutes, à cause d'un événement survenu il y a quelques mois).

Donc il est plus sûr de récupérer la borne supérieure de la boucle via un `select max(id) from deposits`.

Reply:
N'est-ce pas le vrai nombre d'éléments qui nous intéresse plutôt que le `max(id)` ? Étant donné qu'on ne récupère pas tous les éléments mais qu'on a des filtres.
C'est ensuite dans `query.offset(start_index).limit(batch_size).all()` que le `start_index` est utilisé.

Reply:
Oups, oui, pardon. Tu as raison. Je n'avais pas bien fait attention. Mon commentaire n'est valable que si on filtre sur l'id.

(Ceci dit, paginer ainsi via `offset` et `limit` est moins performant que de paginer avec des `where id between x and y`. Mais bon, ce script est du _one-shot_, on survivra.)